test/cli/trailing_commas_test.dart

```diff
@@ -0,0 +1,160 @@
+      await process.shouldExit(64);
+    });
+  });
+
+  test('ignore options file if trailing commas specified on the CLI', () async {
+    await d.dir('foo', [
+      analysisOptionsFile(trailingCommas: TrailingCommas.preserve),
+      d.file('main.dart', _unformatted),
+    ]).create();
+
+    var process = await runFormatterOnDir(['--trailing-commas=automate']);
+    await process.shouldExit(0);
+
+    await d.dir('foo', [d.file('main.dart', _formattedAutomate)]).validate();
+  });
+
+  test('use mode from surrounding options', () async {
+    await _testWithOptions({
+      'formatter': {'trailing_commas': 'preserve'},
+    }, preserve: true);
+  });
+
+  test('use default mode on invalid analysis options', () async {
+    await _testWithOptions({'unrelated': 'stuff'}, preserve: false);
+    await _testWithOptions({'formatter': 'not a map'}, preserve: false);
+    await _testWithOptions({
+      'formatter': {'no': 'trailing_commas'},
+    }, preserve: false);
+    await _testWithOptions({
+      'formatter': {'trailing_commas': 'wombat'},
+    }, preserve: false);
+  });
+
+  test('get mode from included options file', () async {
+    await d.dir('foo', [
+      analysisOptionsFile(include: 'other.yaml'),
+      analysisOptionsFile(name: 'other.yaml', include: 'sub/third.yaml'),
+      d.dir('sub', [
+        analysisOptionsFile(
+          name: 'third.yaml',
+          trailingCommas: TrailingCommas.preserve,
+        ),
+      ]),
+      d.file('main.dart', _unformatted),
+    ]).create();
+
+    var process = await runFormatterOnDir();
+    await process.shouldExit(0);
+
+    // Should preserve trailing commas.
+    await d.dir('foo', [d.file('main.dart', _formattedPreserve)]).validate();
+  });
+
+  group('stdin', () {
+    test('use mode from surrounding package', () async {
+      await d.dir('foo', [
+        analysisOptionsFile(trailingCommas: TrailingCommas.preserve),
+      ]).create();
+
+      var process = await runFormatter(['--stdin-name=foo/main.dart']);
+      process.stdin.writeln(_unformatted);
+      await process.stdin.close();
+
+      // Preserves trailing commas.
+      expect(await process.stdout.next, 'var x = function(');
+      expect(await process.stdout.next, '  argument,');
+      expect(await process.stdout.next, ');');
+      await process.shouldExit(0);
+    });
+
+    test('ignore options file if mode is specified', () async {
+      await d.dir('foo', [
+        analysisOptionsFile(trailingCommas: TrailingCommas.preserve),
+        d.file('main.dart', _unformatted),
+      ]).create();
+
+      var process = await runFormatter([
+        '--trailing-commas=preserve',
+        '--stdin-name=foo/main.dart',
+      ]);
+
+      process.stdin.writeln(_unformatted);
+      await process.stdin.close();
+
+      // Preserves trailing commas.
+      expect(await process.stdout.next, 'var x = function(');
+      expect(await process.stdout.next, '  argument,');
+      expect(await process.stdout.next, ');');
```

Comment:
[optional] It only makes a small difference, but some failure scenarios will have better output if you check the stream queue behavior with expectations instead of only check the values in expectations.

```dart
      await expectLater(process.stdout, emits('var x = function('));
      await expectLater(process.stdout, emits('  argument,'));
      await expectLater(process.stdout, emits(');'));
```

or

```dart
      await expectLater(process.stdout, emitsInOrder([
        'var x = function(',
        '  argument,',
        ');',
      ]);
```

Reply:
Done, here and everywhere.